osquery/tables/system/windows/services.cpp

```diff
@@ -131,32 +153,51 @@ QueryData genServices(QueryContext& context) {
                              &serviceCount,
                              nullptr,
                              nullptr);
+  auto err = GetLastError();
+  if (ERROR_MORE_DATA != err) {
+    return Status(err, "Failed to query service list buffer size");
+  }
 
-  buf = malloc(bytesNeeded);
-  if (EnumServicesStatusEx(schScManager,
-                           SC_ENUM_PROCESS_INFO,
-                           SERVICE_WIN32,
-                           SERVICE_STATE_ALL,
-                           (LPBYTE)buf,
-                           bytesNeeded,
-                           &bytesNeeded,
-                           &serviceCount,
-                           nullptr,
-                           nullptr)) {
-    ENUM_SERVICE_STATUS_PROCESS* services = (ENUM_SERVICE_STATUS_PROCESS*)buf;
-    for (DWORD i = 0; i < serviceCount; ++i) {
-      Row r;
-      if (QuerySvcInfo(schScManager, services[i], r)) {
-        results.push_back(r);
-      }
+  std::unique_ptr<ENUM_SERVICE_STATUS_PROCESS[]> lpSvcBuf(
+      static_cast<ENUM_SERVICE_STATUS_PROCESS*>(malloc(bytesNeeded)));
+  if (lpSvcBuf == nullptr) {
+    return Status(1, "Failed to malloc service buffer");
+  }
+
+  if (0 == EnumServicesStatusEx(scmHandle.get(),
+                                SC_ENUM_PROCESS_INFO,
+                                SERVICE_WIN32,
+                                SERVICE_STATE_ALL,
+                                (LPBYTE)lpSvcBuf.get(),
+                                bytesNeeded,
+                                &bytesNeeded,
+                                &serviceCount,
+                                nullptr,
+                                nullptr)) {
+    return Status(GetLastError(), "Failed to enumerate services");
+  }
+
+  for (size_t i = 0; i < serviceCount; i++) {
+    Row r;
```

Comment:
If you're not adding anything to `Row` here, pass in the `QueryData&` to `getService` and construct a `Row` when you need to within that method.